### DISCUSSION TITLE
What import is more correct?

### DISCUSSION THREAD

**Discussion by adinvadim** [gcanti/fp-ts#1546] [Q&A] [answered]
```
import * as R from 'fp-ts/lib/Record';
import * as A from 'fp-ts/lib/Array';
import { pipe } from 'fp-ts/lib/function'
```

I always import like this. But in issues and few examples see import like (without lib in path)

```
import * as R from 'fp-ts/Record';
import * as A from 'fp-ts/Array';
import { pipe } from 'fp-ts/function'
```

What is more correct? or doesnt matter?

**Comment by pashutk** [comment]
Hi!

Either way is correct. Second approach will work only if you build system is aware what module type should be used (commonjs or es6).

If you want more details you can read about transition between first and second approach in this pull request: https://github.com/gcanti/fp-ts/pull/1241.

**Reply by pashutk** [reply]
It was introduced in 2.8.0 by the way https://github.com/gcanti/fp-ts/releases/tag/2.8.0

**Reply by adinvadim** [reply]
Thank you for link!

**Comment by albohlabs** [comment, marked as answer]
In the`node_modules/fp-ts/Array` folder there is `package.json` with the following content:

```json
{
  "main": "../lib/Array.js",
  "module": "../es6/Array.js",
  "typings": "../lib/Array.d.ts",
  "sideEffects": false
}
```

The important part is the `module` prop which the Rollup docs describes pretty well:

> To make sure your ES modules are immediately usable by tools that work with CommonJS such as Node.js and webpack, you can use Rollup to compile to UMD or CommonJS format, and then point to that compiled version with the main property in your package.json file. If your package.json file also has a module field, ES-module-aware tools like Rollup and webpack 2+ will import the ES module version directly. https://rollupjs.org/guide/en/#publishing-es-modules

This means that the import from `fp-ts/Array` is handled as esm (es-module, ECMAScript module) and comes imo with two major benefits:
* tree-shakeable https://exploringjs.com/es6/ch_modules.html#static-module-structure
* works in the browser https://caniuse.com/es6-module

The `lib` folder contains modules that are in CommonJS format which was the nodejs standard before esm becomes [the official standard format to package JavaScript code](https://tc39.es/ecma262/#sec-modules). 
So the reason this folder exist is to provide backward compatibility.

**Reply by florianbepunkt** [reply]
If I import `import { identity, pipe } from "fp-ts/function";` in an ESM project with TS 4.7, I'll get `Cannot find module 'fp-ts/function' or its corresponding type declarations.`

**Reply by onvlt** [reply]
@florianbepunkt what is your  `module` / `moduleResolution` configuration in `tscofig.json`? I get the same error when `moduleResolution` is `nodenext`, but it works okay with `node`.

**Reply by florianbepunkt** [reply]
Yes. It only happens with `nodenext ` You need to set module resolution to `nodenext` to use ECMA script functionality (see release blog for TS 4.7 rc).

**Reply by qlonik** [reply]
I've been investigating it too a bit. It seems that this approach with lots of `package.json` files in folders doesn't work with `node16`/`nodenext`. It really needs just one top-level package.json with exports map.

**Reply by steida** [reply]
@gcanti Do you plan to release fp-ts 3 with simple package.json? I think a lot of package.json files also break types in Deno https://github.com/esm-dev/esm.sh/issues/300

**Reply by gcanti** [reply]
@steida current v3 (`3.0.0-alpha.22`) has just one top-level package.json with exports map and seems to work fine, so I'd say yes

**Reply by rawnly** [reply]
> @steida current v3 (`3.0.0-alpha.22`) has just one top-level package.json with exports map and seems to work fine, so I'd say yes

@gcanti  Any eta on this? 👀

**Reply by gcanti** [reply]
@Rawnly a top-level package.json with exports map will be in `2.13` https://github.com/gcanti/fp-ts/milestone/60?closed=1

**Comment by maciejw** [comment]
for everyone that want to target nodejs and transpile your program to es6 those imports are illegal

The correct way to do it, is:

```ts
import { option as O } from "fp-ts"
```
doing as docs and examples propose
```ts
import * as O from 'fp-ts/Option';
```
will give you ERR_UNSUPPORTED_DIR_IMPORT, node does not allow importing a subdirectory of a module even if it contains package.json.

 
this assumes also tsconfing.json
```json
{
  "compilerOptions": {
    "module": "NodeNext",
    "moduleResolution": "NodeNext",
    "target": "ES2020",
    "lib": ["ES2020"],
    "sourceMap": true,
    "outDir": "dist",
    "strict": true
  },
  "include": ["src/**/*"]
}

```

and the following package.json

```json
{
  "type": "module",
  "main": "dist/main.js",
  "scripts": {
    "start": "node .",
    "build": "tsc"
  },
  "dependencies": {
    "fp-ts": "^2.16.9"
  },
  "devDependencies": {
    "@types/node": "^22.13.10",
    "typescript": "^5.8.2"
  },
  "name": "fp-ts-tests"
}

```

maybe it will save someone half of a sunday :)
